engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/EngineJobHashingActor.scala

```diff
@@ -0,0 +1,148 @@
+    } else {
+      checkWhetherAllHashesAreKnownAndTransition(newData)
+    }
+  }
+
+  private def checkWhetherAllHashesAreKnownAndTransition(newData: EJHAData) = {
+    if (newData.allHashesKnown) {
+      receiver ! CallCacheHashes(newData.hashesKnown)
+      context.stop(self)
+    }
+    goto(GeneratingAllHashes) using newData
+  }
+
+  /**
+    * Needs to convert a hash result into the set of CachedResults which are consistent with it
+    */
+  private def findRelevantCacheResultsThenTransition(hashResults: Iterable[HashResult]) = {
+    val newData = if (mode.writeToCache) stateData.withNewKnownHashes(hashResults) else stateData
+    self ! CacheResultMatchesForHashes(hashResults, Set.empty) // TODO: Probably want to convert this into a call to another actor which would respond back with the results.
+    stay using newData
+  }
+
+  // TODO: Implement me.
+  // Called at actor start-time. Should convert all job input files into file hash requests
+  private def allInputJobFileHashRequests: JobFileHashRequests = JobFileHashRequests(jobDescriptor.key, List.empty)
+}
+
+object EngineJobHashingActor {
+  trait HashResultMessage {
+    def hashes: Iterable[HashResult]
+  }
+  private[callcaching] case class EJHAInitialHashingResults(hashes: Iterable[HashResult]) extends HashResultMessage
+  private[callcaching] case object CheckWhetherAllHashesAreKnown
+
+  sealed trait EJHAState
+  case object DeterminingHitOrMiss extends EJHAState
+  case object GeneratingAllHashes extends EJHAState
+
+  sealed trait EJHAResponse
+  case class CacheHit(cacheResultId: Int) extends EJHAResponse
+  case object CacheMiss
+  case class CallCacheHashes(hashes: List[HashResult])
+}
+
+/**
+  * Transient data for the EJHA.
+  *
+  * @param possibleCacheResults The set of cache results which have matched all currently tried hashes
+  * @param cacheResultsIntersected The set of hash keys which have already been compared against the cache result table
+  * @param hashesKnown The set of all hashes calculated so far (including initial hashes)
+  * @param hashesNeeded Not transient but oh-so-useful for everything else.
+  */
+private[callcaching] case class EJHAData(possibleCacheResults: Set[Int],
+                                         cacheResultsIntersected: List[HashKey],
+                                         hashesKnown: List[HashResult],
+                                         hashesNeeded: Iterable[HashKey]) {
+
+  // Manipulators
+  def intersectCacheResultIds(newCacheResults: CacheResultMatchesForHashes) = {
+    val newIds = newCacheResults.cacheResultIds
+    val intersectedIds = if (cacheResultsIntersected.nonEmpty) possibleCacheResults.intersect(newIds) else newIds
+    this.copy(
+      possibleCacheResults = intersectedIds,
+      cacheResultsIntersected = cacheResultsIntersected ++ newCacheResults.hashResults.map(_.hashKey))
+  }
+  def withNewKnownHashes(hashResults: Iterable[HashResult]) = this.copy(hashesKnown = hashesKnown ++ hashResults)
```

Comment:
actually...

I almost wonder if `initialiZe` should just return this so it could be assigned to a `val`.  I think what confused me was that a constant is being passed around in the state data.